pkg/extensions/sync/sync.go

```diff
@@ -319,186 +326,227 @@ func syncRegistry(ctx context.Context, regCfg RegistryConfig, upstreamURL string
+					}
+				}
+
+				skipCosignSig, err := canSkipCosignSignature(localRepo, tag, upstreamImageDigest.String(),
+					cosignManifest, imageStore, log)
+				if err != nil {
+					log.Error().Err(err).Msgf("couldn't check if the upstream image %s cosign signature can be skipped",
+						upstreamImageRef.DockerReference())
+				}
 
-			if !skipNotarySig {
-				if err = retry.RetryIfNecessary(ctx, func() error {
-					err = syncNotarySignature(httpClient, imageStore, *registryURL, localRepo, remoteRepo,
-						upstreamImageDigest.String(), refs, log)
+				if !skipCosignSig {
+					if err = retry.RetryIfNecessary(ctx, func() error {
+						err = syncCosignSignature(httpClient, imageStore, *registryURL, localRepo, remoteRepo,
+							upstreamImageDigest.String(), cosignManifest, log)
 
-					return err
-				}, retryOptions); err != nil {
-					log.Error().Err(err).Msgf("couldn't copy notary signature for %s", upstreamImageRef.DockerReference())
+						return err
+					}, retryOptions); err != nil {
+						log.Error().Err(err).Msgf("couldn't copy cosign signature for %s", upstreamImageRef.DockerReference())
+					}
 				}
+
+				continue
 			}
 
-			skipCosignSig, err := canSkipCosignSignature(localRepo, tag, upstreamImageDigest.String(),
-				cosignManifest, imageStore, log)
+			localImageRef, trueCachePath, err := getLocalImageRef(imageStore, localRepo, tag)
 			if err != nil {
-				log.Error().Err(err).Msgf("couldn't check if the upstream image %s cosign signature can be skipped",
-					upstreamImageRef.DockerReference())
+				log.Error().Err(err).Msgf("couldn't obtain a valid image reference for reference %s/%s:%s",
+					localCachePath, localRepo, tag)
+
+				return err
 			}
 
-			if !skipCosignSig {
-				if err = retry.RetryIfNecessary(ctx, func() error {
-					err = syncCosignSignature(httpClient, imageStore, *registryURL, localRepo, remoteRepo,
-						upstreamImageDigest.String(), cosignManifest, log)
+			localCachePath = trueCachePath
 
-					return err
-				}, retryOptions); err != nil {
-					log.Error().Err(err).Msgf("couldn't copy cosign signature for %s", upstreamImageRef.DockerReference())
-				}
-			}
+			log.Info().Msgf("copying image %s to %s", upstreamImageRef.DockerReference(), localCachePath)
 
-			continue
-		}
+			if err = retry.RetryIfNecessary(ctx, func() error {
+				_, err = copy.Image(ctx, policyCtx, localImageRef, upstreamImageRef, &options)
 
-		localImageRef, localCachePath, err := getLocalImageRef(imageStore, localRepo, tag)
-		if err != nil {
-			log.Error().Err(err).Msgf("couldn't obtain a valid image reference for reference %s/%s:%s",
-				localCachePath, localRepo, tag)
+				return err
+			}, retryOptions); err != nil {
+				log.Error().Err(err).Msgf("error while copying image %s to %s",
+					upstreamImageRef.DockerReference(), localCachePath)
 
-			return err
+				return err
+			}
+
+			upstreamImgRefsByTag[tag] = upstreamImageRef
```

Comment:
Now that I think about it, let's pushSyncedLocalImage in this for, no need to do it later in another loop.

The blobs will still be cached in the same temp dir.

This way you don't need to create upstreamImgRefsByTag map, cacheImageStore  and tagsToSync.

Reply:
Moved push logic inside loop